style.md

```diff
@@ -0,0 +1,281 @@
+```julia
+# Bad!
+N = 5
+x = [] #Really bad since it is an Any vector!
+for i in 1:N
+    push!(x, 2.0 * i^2)
+end
+x
+
+# Better!
+x = ones(N)
+for i = 1:N
+    x[i] = 2.0 * i^2
+end
+x
+
+# Best!
+x = [2.0 * i^2 for i in 1:N]
+
+# or
+f(i) = 2.0 * i^2
+x = f.(1:5) #Use broadcasting
+```
+- Prefer `eachindex` to accessing the sizes of matrices.
+```julia
+x = [1.0, 2.0, 5.0, 2.1]
+y = similar(x)
+
+# BAD!
+for i in 1:length(x)
+    y[i] = x[i] + 2
+end
+y
+
+# BETTER!
+for i in eachindex(x)
+    y[i] = x[i] + 2
+end
+y
+
+# GOOD! #No way to preallocate y (although easier ways to write)
+x = rand(10)
+y = similar(x, 0) #Empty of same type as x
+for val in x
+    if val < 0.5
+        push!(y, val)
+    end
+end
+y
+```
+- **Use iterators directly** rather than accessing by index.
+```julia
+x = [1, 2, 3]
+
+# BAD!
+n = 0
+for i in eachindex(x)
+  n += x[i]^2
+end
+n
+
+# BETTER!
+n = 0
+for val in x
+  n += val^2
+end
+n
+
+# BEST!
+sum(xval -> xval^2, x) #i.e. transform each x and then reduce
+```
+
+- **Avoid `range` when possible** and use the `1.0:0.1:1.0` style notation, etc.  Getting rid of the `linspace` is a pain, but use ranges if the exact top doesn't matter.
+```julia
+Base.OneTo(5) # BAD!
+range(1, stop=5) #BAD!
+1:5 #GOOD!
+
+
+N = 10
+zmin = 0.0
+zstop = 1.0
+
+# Old linspace(zmin, zstop, N) doesn't work
+# ACCEPTABLE!
+range(zmin, stop = zstop, length = N)
+```    
+  - Furthermore, if you don't really care if if hits the `zstop` exactly and are willing to give a stepsize then, the following is the clearest
+```julia
+zmin = 0.0
+zstop = 1.0
+step = 0.1
+
+# GOOD! But..
+r = zmin:step:zstop
+
+# ...CAREFUL
+r = 0.0:0.22:1.0 # Note the end isn't a multiple of the step...
+@assert r == 0.0:0.22:0.88
+@assert maximum(r) == 0.88 #Use to get the maxium of the range, perhaps != 
+```
+
+
+
+## Dependencies
+
+- **Use external packages** whenever possible, and never rewrite code that is available in a well-maintained external package (even if it is imperfect)
+- The following packages can be used as a dependency without any concerns: `QuantEcon, Parameters, Optim, Roots, Expectations, NLsolve, DataFrames, LaTeXStrings, Plots, ....`
+- **Do** use `using` where possible (i.e. not `import`), and include the whole package as opposed to selecting only particular functions or types.
+- **Prefer** to keep packages used throughout the lecture at the top of the first block (e.g. `using LinearAlgebra, Parameters, QuantEcon`)  but packages used only in a single place should have the `using` local to that use.
+    - If `Plots` is only used lower down in the lecture, then try to have it local to that section to ensure faster loading time.
+- **Always seed random numbers** in order for automated testing to function.
+
+## Work in Progress Discussions
+1. When and how do we use `eachindex` with multidimensional arrays you want to iterate through?
```

Comment:
```
for i in 1:size(X, 1)
    for j in 1:size(X, 2)
    end
end
```
Always forget which one goes first, but it is in the docs.

Reply:
Is there a way to do it cleanly with `eachindex` or something similar, or is that the preferred way?

Reply:
@Nosferican rule of thumb is to always make the early index the inner most loop.

So with `x = rand(2, 2, 2)` it would be most efficient to iterate `for k in 1:size(x, 3), j in 1:size(x, 2), i in 1:size(x, 1)`

Reply:
`CartesianIndices`, but probably not something taught at beginner's level. If possible a helper and `mapreduce` or a nested `for` loop.

Reply:
> So with x = rand(2, 2, 2) it would be most efficient to iterate for k in 1:size(x, 3), j in 1:size(x, 2), i in 1:size(x, 1) But as hinted here, eachindex will take care of that for us!

OK, I think I have added something there now.  But @sglyon  How exactly do you do it with `eachindex` in two dimensions, and still get access to the indices themselves?  If there is no clean way to do it, we can just use the julia style guide fallback and say nothing.

Reply:
```
X = rand(1,2,2);
for idx ∈ CartesianIndices(X)
       println(idx)
       end
CartesianIndex(1, 1, 1)
CartesianIndex(1, 2, 1)
CartesianIndex(1, 1, 2)
CartesianIndex(1, 2, 2)
```